src/Microsoft.OpenApi.Hidi/Formatters/PowerShellFormatter.cs

```diff
@@ -0,0 +1,261 @@
+        {
+            var charPos = operationId.LastIndexOf('.', operationId.Length - 1);
+            if (operationId.Contains('_') || charPos < 0)
+                return operationId;
+            var newOperationId = new StringBuilder(operationId);
+            newOperationId[charPos] = '_';
+            operationId = newOperationId.ToString();
+            return operationId;
+        }
+
+        private static string ResolvePutOperationId(string operationId)
+        {
+            return operationId.Contains(DefaultPutPrefix) ?
+                operationId.Replace(DefaultPutPrefix, PowerShellPutPrefix) : operationId;
+        }
+
+        private static string ResolveByRefOperationId(string operationId)
+        {
+            // Update $ref path operationId name
+            // Ref key word is enclosed between lower-cased and upper-cased letters
+            // Ex.: applications_GetRefCreatedOnBehalfOf to applications_GetCreatedOnBehalfOfByRef
+            return s_oDataRefRegex.Match(operationId).Success ? $"{s_oDataRefRegex.Replace(operationId, string.Empty)}ByRef" : operationId;
+        }
+
+        private static string ResolveODataCastOperationId(string operationId)
+        {
+            var match = s_oDataCastRegex.Match(operationId);
+            return match.Success ? $"{match.Groups[1]}{match.Groups[2]}" : operationId;
+        }
+
+        private static string SingularizeAndDeduplicateOperationId(IList<string> operationIdSegments)
+        {
+            var segmentsCount = operationIdSegments.Count;
+            var lastSegmentIndex = segmentsCount - 1;
+            var singularizedSegments = new List<string>();
+
+            for (int x = 0; x < segmentsCount; x++)
+            {
+                var segment = operationIdSegments[x].Singularize(inputIsKnownToBePlural: false);
+
+                // If a segment name is contained in the previous segment, the latter is considered a duplicate.
+                // The last segment is ignored as a rule.
+                if ((x > 0 && x < lastSegmentIndex) && singularizedSegments[singularizedSegments.Count - 1].Equals(segment, StringComparison.OrdinalIgnoreCase))
+                    continue;
+
+                singularizedSegments.Add(segment);
+            }
+            return string.Join(".", singularizedSegments);
+        }
+
+        private static string RemoveHashSuffix(string operationId)
+        {
+            // Remove hash suffix values from OperationIds.
+            return s_hashSuffixRegex.Match(operationId).Value;
+        }
+
+        private static string RemoveKeyTypeSegment(string operationId, IList<OpenApiParameter> parameters)
+        {
+            var segments = operationId.SplitByChar('.');
+            foreach (var parameter in parameters)
+            {
+                var keyTypeExtension = parameter.Extensions.GetExtension("x-ms-docs-key-type");
+                if (keyTypeExtension != null && operationId.Contains(keyTypeExtension))
```

Comment:
contains is going to be slow on a list, have we considered a hashset instead?

Reply:
The contains is on a string, operationId. We are looking at O(n) as the worst case (left -> right search).

Reply:
right, sorry about that. Should we at least specify StringComparison?

Reply:
Good call! Let me address that.